src/style-spec/expression/definitions/within.js

```diff
@@ -0,0 +1,273 @@
+    const pointBBox = [Infinity, Infinity, -Infinity, -Infinity];
+    const lngLatPoints = [];
+    for (const points of ctx.geometry()) {
+        for (const point of points) {
+            const p = getLngLatPoint(point, ctx.canonicalID());
+            lngLatPoints.push(p);
+            updateBBox(pointBBox, p);
+        }
+    }
+    if (!boxWithinBox(pointBBox, polyBBox)) return false;
+    for (let i = 0; i < lngLatPoints.length; ++i) {
+        if (!pointWithinPolygons(lngLatPoints[i], polygonGeometry)) return false;
+    }
+    return true;
+}
+
+function linesWithinPolygons(ctx: EvaluationContext, polygonGeometry: GeoJSONPolygons, polyBBox: BBox) {
+    const lineBBox = [Infinity, Infinity, -Infinity, -Infinity];
+    const lineCoords = [];
+    for (const line of ctx.geometry()) {
+        const lineCoord = getLngLatPoints(line, ctx.canonicalID());
+        lineCoords.push(lineCoord);
+        calcBBox(lineBBox, lineCoord, 'LineString');
+    }
+    if (!boxWithinBox(lineBBox, polyBBox)) return false;
+    for (let i = 0; i < lineCoords.length; ++i) {
+        if (!lineStringWithinPolygons(lineCoords[i], polygonGeometry)) return false;
+    }
+    return true;
+}
+
+class Within implements Expression {
+    type: Type;
+    geojson: GeoJSON
+    geometries: GeoJSONPolygons;
+    polyBBox: BBox;
+
+    constructor(geojson: GeoJSON, geometries: GeoJSONPolygons) {
+        this.type = BooleanType;
+        this.geojson = geojson;
+        this.geometries = geometries;
+        this.polyBBox = [Infinity, Infinity, -Infinity, -Infinity];
+        calcBBox(this.polyBBox, this.geometries.coordinates, this.geometries.type);
+    }
+
+    static parse(args: $ReadOnlyArray<mixed>, context: ParsingContext) {
+        if (args.length !== 2)
+            return context.error(`'within' expression requires exactly one argument, but found ${args.length - 1} instead.`);
+        if (isValue(args[1])) {
+            const geojson = (args[1]: Object);
+            if (geojson.type === 'FeatureCollection') {
+                for (let i = 0; i < geojson.features.length; ++i) {
+                    const type = geojson.features[i].geometry.type;
+                    if (type === 'Polygon' || type === 'MultiPolygon') {
+                        return new Within(geojson, geojson.features[i].geometry);
+                    }
+                }
+            } else if (geojson.type === 'Feature') {
+                const type = geojson.geometry.type;
+                if (type === 'Polygon' || type === 'MultiPolygon') {
+                    return new Within(geojson, geojson.geometry);
+                }
+            } else if (geojson.type  === 'Polygon' || geojson.type === 'MultiPolygon') {
+                return new Within(geojson, geojson);
+            }
+        }
+        return context.error(`'within' expression requires valid geojson source that contains polygon geometry type.`);
```

Comment:
if this doesn't take an actual source object as an argument as @tristen suggested in another comment, then using the word `source` here may be confusing since it has a specific meaning in the context of GL JS